### ISSUE TITLE
Duplicate stream item ids

### ISSUE THREAD

**Issue by RomanKotov** [phoenixframework/phoenix_live_view#2689]
### Environment

* Elixir version (elixir -v): `1.14.4-otp-25`
* Phoenix version (mix deps): `1.7.2`
* Phoenix LiveView version (mix deps): `0.19.2` and master branch https://github.com/phoenixframework/phoenix_live_view/commit/7aa516ff8107196a27cc11982415d86958fc8b98
* Operating system: macOS
* Browsers you attempted to reproduce this bug on (the more the merrier): Safari, Firefox
* Does the problem persist after removing "assets/node_modules" and trying again? Yes/no: Yes

### Actual behavior

I render batch updates for the items.
An item with the same id can occur multiple times in a single batch.

If I run `stream_insert/4` twice, with the same id, I will get two HTML elements with the same id.
I will get the same result, if I use insert all batch items at once with `stream/4`.

### Expected behavior

Ids should be unique in the markup.

I see following options about how it could work:
1. Show only the latest version of the item.
2. Raise an error for duplicate ids.
3. Document this behaviour and propose a workaround for it.

I am not sure which option fits best here.

### Way to reproduce:

I have created a `mix install` example for this issue

```elixir
Application.put_env(:sample, Example.Endpoint,
  http: [ip: {127, 0, 0, 1}, port: 5001],
  server: true,
  live_view: [signing_salt: "aaaaaaaa"],
  secret_key_base: String.duplicate("a", 64)
)

Mix.install([
  {:plug_cowboy, "~> 2.5"},
  {:jason, "~> 1.0"},
  {:phoenix, "~> 1.7.2", override: true},
  {:phoenix_live_view, "~> 0.19.2"}
  # {:phoenix_live_view, github: "phoenixframework/phoenix_live_view", ref: "7aa516f"}
])

defmodule Example.ErrorView do
  def render(template, _), do: Phoenix.Controller.status_message_from_template(template)
end

defmodule Example.HomeLive do
  use Phoenix.LiveView, layout: {__MODULE__, :live}

  def mount(_params, _session, socket) do
    socket =
      socket
      |> assign(:count, 0)
      |> stream(:items, [])

    {:ok, socket}
  end

  def render("live.html", assigns) do
    ~H"""
    <script src="https://cdn.jsdelivr.net/npm/phoenix@1.7.2/priv/static/phoenix.min.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/phoenix_live_view@0.19.2/priv/static/phoenix_live_view.min.js"></script>
    <script>
      let liveSocket = new window.LiveView.LiveSocket("/live", window.Phoenix.Socket)
      liveSocket.connect()
    </script>
    <style>
      * { font-size: 1.1em; }
    </style>
    <%= @inner_content %>
    """
  end

  def render(assigns) do
    ~H"""
    <div> total items: <%= @count %> </div>
    <button phx-click="add_item">add item</button>
    <button phx-click="add_item_batch">add item batch</button>
    <ul id="items" phx-update="stream">
      <li id={id} :for={{id, item} <- @streams.items}>
        <%= inspect(item) %>
      </li>
    </ul>
    """
  end

  def handle_event("add_item", _params, socket) do
    count = socket.assigns.count

    item = %{id: count, name: "single_original"}
    changed_item = %{id: count, name: "changed_original"}

    socket =
      socket
      |> assign(:count, count + 1)
      |> stream_insert(:items, item)
      |> stream_insert(:items, changed_item)

    {:noreply, socket}
  end

  def handle_event("add_item_batch", _params, socket) do
    count = socket.assigns.count

    item = %{id: count, name: "batch_original"}
    changed_item = %{id: count, name: "batch_changed"}

    socket =
      socket
      |> assign(:count, count + 1)
      |> stream(:items, [item, changed_item])

    {:noreply, socket}
  end
end

defmodule Example.Router do
  use Phoenix.Router
  import Phoenix.LiveView.Router

  pipeline :browser do
    plug(:accepts, ["html"])
  end

  scope "/", Example do
    pipe_through(:browser)

    live("/", HomeLive, :index)
  end
end

defmodule Example.Endpoint do
  use Phoenix.Endpoint, otp_app: :sample
  socket("/live", Phoenix.LiveView.Socket)
  plug(Example.Router)
end

{:ok, _} = Supervisor.start_link([Example.Endpoint], strategy: :one_for_one)
Process.sleep(:infinity)
```

Here is example screenshot:
![Screenshot 2023-06-13 at 16 21 50](https://github.com/phoenixframework/phoenix_live_view/assets/10118630/3c6dc755-7b08-412c-acf8-b854a03172ff)

### Workaround

Now I deduplicate batch items, so item with the same ID occurs only once in the batch.

**Comment by puruzio** [comment]
I'm having a similar issue. If I do "view source page" in a browser for a page that contains a stream containing 3 items (ie. 3 li tags), the source code shows the 3 items repeated twice (ie. 6 li tags), and the browser console complains about it. If I add an item to the stream (by resetting the stream with a 4 item list, the browser briefly shows 8 items and quickly hides the duplicated ones leaving the intended 4.

**Comment by Malian** [comment]
When I was trying to find workarounds for https://github.com/phoenixframework/phoenix_live_view/issues/2826, I also noticed duplicated stream item when `stream` is called with `reset: true` twice.

```elixir
socket
|> stream(:items, items, reset: true)
|> stream(:items, items, reset: true)
```

If you believe I should open a new issue, I would be happy to do it with a simple app that reproduces the bug.

**Comment by chrismccord** [comment]
This is expected. You'll need to dedup your items in your insert calls. Thanks!

**Comment by rapidfsub** [comment]
@chrismccord 
Excuse me for jumping in on this older thread, but could you explain about the reason why this is expected?

**Comment by RomanKotov** [comment]
> @chrismccord Excuse me for jumping in on this older thread, but could you explain about the reason why this is expected?

Hi @rapidfsub ,

I have looked into the implementation, and there is no deduplication for stream items (at least for now).

LiveView operates in batches. Batch starts with an event. For example, user presses a button and Phoenix calls our `handle_event/3` callback. We modify a socket (for example, assign items) and return it back to Phoenix. Phoenix computes diff and sends it over the wire. Batch is finished.

LiveView streams work in batches too. They collect all changes since the start of batch. LiveView clears them after the diff was rendered:
https://github.com/phoenixframework/phoenix_live_view/blob/c7b12df05a3e27568fa07221752071ac92fd6dd4/lib/phoenix_live_view.ex#L1882

This means, that streams **accumulate all changes** during the batch.

Each time you add a stream item, it gets **appended to the end** of `inserts`:
https://github.com/phoenixframework/phoenix_live_view/blob/c7b12df05a3e27568fa07221752071ac92fd6dd4/lib/phoenix_live_view/live_stream.ex#L64

If you add a `reset: true` option, it will update only the `reset` field of a stream struct. It is a signal for Browser part to drop previous contents of stream-related container. It does nothing to the previous items in batch:
https://github.com/phoenixframework/phoenix_live_view/blob/c7b12df05a3e27568fa07221752071ac92fd6dd4/lib/phoenix_live_view/live_stream.ex#L46

If you assign multiple items via `stream`, they will be appended to the end of previous inserts, one item at a time:
https://github.com/phoenixframework/phoenix_live_view/blob/c7b12df05a3e27568fa07221752071ac92fd6dd4/lib/phoenix_live_view.ex#L1856

HTML allows multiple items with the same `id` on the page. Ids are meant to be different so `document.getElementById` will take only the first item. But browser will be happy to render multiple items with the same id.

It seems that `live_view` browser integration believes that all the diffs from server are correct. It will add every item, if server tells that it should be added (even if they have the same id). Browser will not complain about it too.

### Conclusion
- LiveView streams do not deduplicate items by ids. **Programmers should do this themselves**.
- LiveView resets streams after render (batch). Streams accumulate changes during a batch.
- `reset: true` does not drop previously accumulated stream items. It only signals browser to clear old data from stream-related container.
- Deduplicating items in a LiveView would decrease performance of the application. If we want to add an item, we should deduplicate a list after each insert. If you need to add N items, you will need to deduplicate list N times too (roughly N<sup>2</sup> operations).

### Possible improvements
- We can mention this caveat in the LiveView documentation.
- **Prepend the item to the beginning of `LiveStream.inserts`** and reverse this list before render, [instead of appending it to the end of the list](https://github.com/phoenixframework/phoenix_live_view/blob/c7b12df05a3e27568fa07221752071ac92fd6dd4/lib/phoenix_live_view/live_stream.ex#L64). Prepending is O(1) operation. Appending copies a list every time, so it is O(N) one. This is also [Erlang performance recommendation](https://erlang.org/documentation/doc-15.0-rc3/doc/system/listhandling.html#creating-a-list). It can optimize LiveView stream performance for large number of items.

**Comment by RomanKotov** [comment]
I can prepare PR (or PRs) with "possible improvements", if it is useful.

**Comment by rapidfsub** [comment]
@RomanKotov, I was curious about the reason behind the issue. Thank you for your explanation.

I have identified the following pain point scenarios:

Scenario 1:

1. Call stream/4 in the mount callback.
2. Call stream/4 in the handle_params callback.

Scenario 2:

1. Call stream/4 in the handle_event callback.
2. Perform a push_patch in the handle_event callback.
3. Call stream/4 in the handle_params callback.

I want each callback not to know each other's implementation,
but I can't because there is no dedup.

**Comment by rapidfsub** [comment]
But I understood by your explanation.
If LiveView Stream always do dedup automatically, then there would be a performance issue with large amount of inserts.
So I'll dedup by myself.
